contracts/src/FunctionGateway.sol

```diff
@@ -28,174 +29,200 @@ contract FunctionGateway is IFunctionGateway, FunctionRegistry, TimelockedUpgrad
+        (bool status, ) = _callbackAddress.call(
+            abi.encodeWithSelector(_callbackSelector, _output, _context)
+        );
+        if (!status) {
+            revert CallbackFailed(_callbackAddress, _callbackSelector);
         }
 
-        emit ProofBatchFulfilled(
-            _requestIds, _aggregateProof, _inputsRoot, _outputHashes, _outputsRoot, _verificationKeyRoot
-        );
+        emit CallbackFulfilled(_nonce, _input, _output);
     }
 
-    /// @dev Fulfills the callback for a request.
+    /// @dev zkCall is like `call`
     /// @param _requestId The id of the request to be fulfilled.
     /// @param _output The output of the proof.
     /// @param _context The context of the runtime.
-    function callback(bytes32 _requestId, bytes memory _output, bytes memory _context) external {
-        // Do some sanity checks.
-        FunctionRequest storage r = requests[_requestId];
-        if (r.callbackFulfilled) {
-            revert CallbackAlreadyFulfilled(_requestId);
-        } else if (r.callbackAddress == address(0)) {
-            revert RequestNotFound(_requestId);
-        } else if (r.contextHash != keccak256(_context)) {
-            revert ContextMismatch(_requestId, _context);
-        } else if (r.outputHash != sha256(_output)) {
-            revert OutputMismatch(_requestId, _output);
-        } else if (!r.proofFulfilled) {
-            revert ProofNotFulfilled(_requestId);
+    function zkCall(
+        bytes32 _functionId,
+        bytes memory _input
+    ) external returns (bool, bytes memory) {
+        if (
+            currentVerifiedCall.functionId == _functionId &&
+            currentVerifiedCall.inputHash == sha256(_input)
+        ) {
+            return (true, currentVerifiedCall.output);
+        } else {
+            // TODO: process payment
+            emit ProofRequested(_functionId, _input);
+            return (false, "");
+        }
+    }
+
+    function zkCallFulfill(
+        bytes32 _functionId,
+        bytes memory _input,
+        bytes memory _output,
+        bytes memory _proof,
+        address _callbackAddress,
+        bytes memory _callbackData
+    ) external {
+        bytes32 inputHash = sha256(_input);
+        bytes32 outputHash = sha256(_output);
+
+        // Verify the proof.
+        address verifier = verifiers[r.functionId];
+        if (
+            !IFunctionVerifier(verifier).verify(inputHash, outputHash, _proof)
+        ) {
+            revert InvalidProof(
+                address(verifier),
+                inputHash,
+                _outputHash,
+                _proof
+            );
         }
 
-        // Update the request.
-        r.callbackFulfilled = true;
+        if (_callbackAddress == address(0)) {
+            // If there is no callback, then just store the result
+            verifiedResults[functionId][inputHash] = outputHash;
+            return;
+        }
+
+        VerifiedCall currentVerifiedCall = VerifiedCall({
```

Comment:
I'm really surprised if saving `currentVerifiedCall` to memory like this works, pretty sure it needs to be a storage slot.